### PR TITLE
Price Floors: Implement enforcement

### PR DESCRIPTION
**Pull request by rpanchyk**
The enforcement is about to accept or reject the response bids according to previously defined bid floors.

### CODE REVIEW

File: src/test/java/org/prebid/server/floors/BasicPriceFloorEnforcerTest.java [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+        // when
+        final AuctionParticipation result = enforcer.enforce(auctionParticipation, account);
+
+        // then
+        assertThat(singleton(result))
+                .extracting(AuctionParticipation::getBidderResponse)
+                .extracting(BidderResponse::getSeatBid)
+                .flatExtracting(BidderSeatBid::getBids, BidderSeatBid::getErrors)
+                .containsExactly(
+                        List.of(BidderBid.of(
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/floors/BasicPriceFloorEnforcerTest.java:238]: For one object let's use `singletonList` same for 239 line

**Review by snahornyi — CHANGES_REQUESTED** [review]

File: src/test/java/org/prebid/server/auction/ExchangeServiceTest.java [hunk 2]
```diff
@@ -3234,6 +3238,60 @@ public void shouldReturnBidsWithoutAdjustingPricesWhenAdjustmentFactorNotPresent
...
+        // given
+        givenBidder("bidder1", mock(Bidder.class), givenSeatBid(singletonList(
+                givenBid(Bid.builder().id("bidId1").impid("impId1").build(), "USD"))));
+
+        final BidRequest bidRequest = givenBidRequest(singletonList(
+                        // imp ids are not really used for matching, included them here for clarity
+                        givenImp(singletonMap("bidder1", 1), builder -> builder.id("impId1"))),
+                identity());
+
+        given(priceFloorEnforcer.enforce(any(), any())).willAnswer(inv -> inv.getArgument(0));
```
[the remark this answers is not in the document]
Comment (rpanchyk) [reply src/test/java/org/prebid/server/auction/ExchangeServiceTest.java:3281]: Just wanted to emphasize, but ok -)

File: src/test/java/org/prebid/server/auction/ExchangeServiceTest.java [hunk 3]
```diff
@@ -3234,6 +3238,60 @@ public void shouldReturnBidsWithoutAdjustingPricesWhenAdjustmentFactorNotPresent
                 .containsExactly(BigDecimal.ONE);
     }
 
+    @Test
+    public void shouldReturnNoBidsIfPriceFloorEnforcerRemoveThem() {
```
[the remark this answers is not in the document]
Comment (rpanchyk) [reply src/test/java/org/prebid/server/auction/ExchangeServiceTest.java:3242]: This case is in `BasicPriceFloorEnforcerTest`, would you to repeat it here?

File: src/main/java/org/prebid/server/spring/config/PriceFloorsConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package org.prebid.server.spring.config;
+
+import org.prebid.server.floors.BasicPriceFloorEnforcer;
+import org.prebid.server.floors.PriceFloorEnforcer;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class PriceFloorsConfiguration {
```
[the remark this answers is not in the document]
Comment (rpanchyk) [reply src/main/java/org/prebid/server/spring/config/PriceFloorsConfiguration.java:10]: Just suppose we'll have fetcher and signalling beans here. So, probably separate place for feature services is better choice.

File: src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+    private static boolean shouldApplyEnforcement(AuctionParticipation auctionParticipation,
+                                                  AccountPriceFloorsConfig accountPriceFloorsConfig) {
+
+        if (accountPriceFloorsConfig == null || BooleanUtils.isNotTrue(accountPriceFloorsConfig.getEnabled())) {
+            return false;
+        }
+
+        final BidderRequest bidderRequest = auctionParticipation.getBidderRequest();
+        final BidRequest bidderBidRequest = ObjectUtil.getIfNotNull(bidderRequest, BidderRequest::getBidRequest);
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java:64]: Is it possible null bidderRequest?

File: src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+        final ExtRequestPrebidFloors floors = getFloors(bidderBidRequest);
+        final ExtRequestPrebidFloorsEnforcement enforcement = ObjectUtil.getIfNotNull(floors,
+                ExtRequestPrebidFloors::getEnforcement);
+
+        return isEnforcedByRequest(enforcement) && isSatisfiedByEnforceRate(enforcement);
+    }
+
+    private static ExtRequestPrebidFloors getFloors(BidRequest bidRequest) {
+        final ExtRequest ext = ObjectUtil.getIfNotNull(bidRequest, BidRequest::getExt);
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java:73]: Is it possible null bidRequest?

File: src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+                    resolvedFloorCurrency,
+                    bidRequestCurrency);
+        }
+
+        return floor;
+    }
+
+    private static String resolveBidRequestCurrency(BidRequest bidRequest) {
+        final List<String> currencies = bidRequest.getCur();
+        return CollectionUtils.isEmpty(currencies) ? null : currencies.get(0);
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java:197]: Nitpick. \n before return. Also pls check similar

File: src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+                ExtRequestPrebidFloorsEnforcement::getEnforcePbs);
+
+        return BooleanUtils.isNotFalse(enforcePbs);
+    }
+
+    private static boolean isSatisfiedByEnforceRate(ExtRequestPrebidFloorsEnforcement enforcement) {
+        final Integer enforceRate = ObjectUtil.getIfNotNull(enforcement,
+                ExtRequestPrebidFloorsEnforcement::getEnforceRate);
+
+        return enforceRate == null || (enforceRate >= 0 && enforceRate <= 100
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java:90]: False result in this condition(enforceRate >= 0 && enforceRate <= 100) is config mistake. Should we log it?

File: src/test/java/org/prebid/server/auction/ExchangeServiceTest.java [hunk 9]
```diff
@@ -3234,6 +3239,38 @@ public void shouldReturnBidsWithoutAdjustingPricesWhenAdjustmentFactorNotPresent
                 .containsExactly(BigDecimal.ONE);
     }
 
+    @Test
+    public void shouldReturnBidsAcceptedByPriceFloorEnforcer() {
+        // given
+        final BidderBid bid1 = givenBid(Bid.builder().id("bidId1").impid("impId1").price(ONE).build(), "USD");
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/auction/ExchangeServiceTest.java:3245]: Usually we use firstBid. But in this situation should we highlight in bid name that bid should be enforced, like bidToEnforce or something similar?

File: src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+                                         BidRequest bidderBidRequest,
+                                         BidRequest bidRequest) {
+
+        final String resolvedFloorCurrency = ObjectUtils.defaultIfNull(floorCurrency,
+                resolveBidRequestCurrency(bidderBidRequest));
+
+        final String bidRequestCurrency = resolveBidRequestCurrency(bidRequest);
+
+        if (resolvedFloorCurrency != null && !resolvedFloorCurrency.equals(bidRequestCurrency)) {
+            return currencyConversionService.convertCurrency(
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java:185]: Should we handle exceptions from this method?

File: src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+    }
+
+    private static Imp correspondingImp(Bid bid, List<Imp> imps) {
+        final String impId = bid.getImpid();
+        return ListUtils.emptyIfNull(imps).stream()
+                .filter(imp -> Objects.equals(impId, imp.getId()))
+                .findFirst()
+                // Should never happen, see ResponseBidValidator usage.
+                .orElseThrow(() -> new PreBidException(
+                        String.format("Bid with impId %s doesn't have matched imp", impId)));
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java:207]: This exception should be promoted, or handled inside this class?

**Review by snahornyi — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java (hunk 5 above)
Comment (rpanchyk) [reply src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java:64]: Seems null is covered.

File: src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java (hunk 6 above)
Comment (rpanchyk) [reply src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java:73]: Seems null is covered.

File: src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java (hunk 8 above)
Comment (rpanchyk) [reply src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java:90]: In this case `enforceRate` is part of bid request, not account config.

File: src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java (hunk 10 above)
Comment (rpanchyk) [reply src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java:185]: Good catch, thx!

File: src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java (hunk 11 above)
Comment (rpanchyk) [reply src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java:207]: i'd say seems validation is performed earlier by PBS-Core - let's just throw it upper.

File: src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java (hunk 7 above)
Comment (snahornyi) [reply src/main/java/org/prebid/server/floors/BasicPriceFloorEnforcer.java:197]: But return it is more like separate part which shows what will be given outside the method

**Review by snahornyi — APPROVED** [review]

**Pull request merged**